toolchain/base/int_store.h

```diff
@@ -0,0 +1,186 @@
+    if (IntId id = IntId::TryMakeUnsignedValue(value); id.is_valid())
+        [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return AddUnsignedLarge(std::move(value));
+  }
+
+  // Returns the value for an ID.
+  //
+  // This will always be a signed `APInt` with a canonical bit width for the
+  // specific integer value in question.
+  auto Get(IntId id) const -> llvm::APInt {
+    if (id.is_value()) [[likely]] {
+      return llvm::APInt(MinAPWidth, id.AsValue(), /*isSigned=*/true);
+    }
+    return values_.Get(APIntId(id.AsIndex()));
+  }
+
+  // Returns the value for an ID adjusted to a specific bit width.
+  //
+  // Note that because we store canonical mathematical integers as signed
+  // integers, this always sign extends or truncates to the target width. The
+  // caller can then use that as a signed or unsigned integer as needed.
+  auto GetAtWidth(IntId id, int bit_width) const -> llvm::APInt {
+    llvm::APInt value = Get(id);
+    if (static_cast<int>(value.getBitWidth()) != bit_width) {
+      value = value.sextOrTrunc(bit_width);
+    }
+    return value;
+  }
+
+  // Returns the value for an ID adjusted to the bit width specified with
+  // another integer ID.
+  //
+  // This simply looks up the width integer ID, and then calls the above
+  // `GetAtWidth` overload using the value found for it. See that overload for
+  // more details.
+  auto GetAtWidth(IntId id, IntId bit_width_id) const -> llvm::APInt {
+    const llvm::APInt& bit_width = Get(bit_width_id);
+    CARBON_CHECK(bit_width.isStrictlyPositive() &&
+                     bit_width.isSignedIntN(sizeof(int) * 8),
+                 "Invalid bit width value: {0}", bit_width);
+    return GetAtWidth(id, bit_width.getSExtValue());
+  }
+
+  // Looks up the canonical ID for a value, or returns invalid if not in the
+  // store.
+  auto LookupSigned(llvm::APInt value) const -> IntId {
+    if (IntId id = IntId::TryMakeSignedValue(value); id.is_valid()) [[likely]] {
+      return id;
+    }
+
+    // Fallback for larger values.
+    return LookupSignedLarge(std::move(value));
+  }
+
+  // Output a YAML description of this data structure. Note that this will only
+  // include the integers that required storing, not those successfully embedded
+  // into the ID space.
+  auto OutputYaml() const -> Yaml::OutputMapping;
+
+  auto array_ref() const -> llvm::ArrayRef<llvm::APInt> {
+    return values_.array_ref();
+  }
+  auto size() const -> size_t { return values_.size(); }
+
+  // Collects the memory usage of the separately stored integers.
+  auto CollectMemUsage(MemUsage& mem_usage, llvm::StringRef label) const
+      -> void;
+
+ private:
+  friend struct Testing::IntStoreTestPeer;
+
+  struct APIntId : IdBase, Printable<APIntId> {
+    using ValueType = llvm::APInt;
+    static const APIntId Invalid;
+    using IdBase::IdBase;
+    auto Print(llvm::raw_ostream& out) const -> void {
+      out << "ap-int";
```

Comment:
```suggestion
      out << "ap_int";
```

Looking at check/ids.h, we have things like `check_ir` (underscore) rather than hyphenation.